### PR TITLE
feat Add parsing of arithmetics

### PR DESCRIPTION
**Pull request by prsabahrami**
https://github.com/prefix-dev/shell/pull/134#issuecomment-2357466150

### CODE REVIEW

**Comment by prsabahrami** [comment]
@wolfv @certik Let me know how you think about the way AST is being created here. Then I can proceed with the execution.

**Comment by wolfv** [comment]
To me, this looks great!

**Comment by prsabahrami** [comment]
Great! Then I'll proceed with execution and take care of the unwraps.

**Comment by prsabahrami** [comment]
This seems to be working. I have added support for almost everything except `++` and `--` operators. Those are also easy to implement. Will take care of them afterwards.

File: crates/deno_task_shell/Cargo.toml [hunk 1]
```diff
@@ -29,6 +29,7 @@ pest_derive = "2.7.12"
 dirs = "5.0.1"
 pest_ascii_tree = { git = "https://github.com/prsabahrami/pest_ascii_tree.git", branch = "master" }
 miette = "7.2.0"
+lazy_static = "1.4.0"
```
Comment (Hofer-Julian) [inline crates/deno_task_shell/Cargo.toml:32]: I don't see this dependency used anywhere :)

File: crates/deno_task_shell/src/shell/command.rs [hunk 2]
```diff
@@ -194,19 +194,23 @@ async fn parse_shebang_args(
...
-  super::execute::evaluate_args(
+  let result = super::execute::evaluate_args(
     cmd.args,
     &context.state,
     context.stdin.clone(),
     context.stderr.clone(),
   )
   .await
-  .map_err(|e| miette!(e.to_string()))
+  .map_err(|e| miette!(e.to_string()))?;
```
Comment (Hofer-Julian) [inline crates/deno_task_shell/src/shell/command.rs:212]:
```suggestion
  .into_diagnostics()?;
```

Does this work as well?

**Review by Hofer-Julian — COMMENTED** [review]
A few minor comments

File: crates/deno_task_shell/Cargo.toml (hunk 1 above)
Comment (prsabahrami) [reply crates/deno_task_shell/Cargo.toml:32]: It is imported in parser.rs and used in line 533 to define precedence over ops.

**Comment by wolfv** [comment]
I found like one little thing when testing.

```
A=5
echo $((1+A))
# doesn't work
echo $((1+$A))
echo $((1+${A}))
```

Not sure how hard that is. Not super important IMO!

**Comment by prsabahrami** [comment]
> I found like one little thing when testing.
> 
> 
> 
> ```
> 
> A=5
> 
> echo $((1+A))
> 
> # doesn't work
> 
> echo $((1+$A))
> 
> echo $((1+${A}))
> 
> ```
> 
> 
> 
> Not sure how hard that is. Not super important IMO! 

Thanks for pointing this out. If the behaviour in such cases is going to be the same as the first working example, then handling these would a slight modification to the parser if I'm not mistaken.

**Comment by wolfv** [comment]
yep, definitely the same. Although with the `${A:1:2}` syntax you could do additional things in the future (cf https://github.com/prefix-dev/shell/pull/148).

And then in the bash case, things like `$((${A}1+1))` would also work but are probably annoying to support.

```
╭─wolfv@pixi ~/Programs/shell ‹support_arithmetic●›
╰─$ echo $A                                                                                                                                                                                                                                             
5
╭─wolfv@pixi ~/Programs/shell ‹support_arithmetic●›
╰─$ echo $((${A}5+1))
56
```

**Comment by prsabahrami** [comment]
Cool. I'll look into those and add support for. But, I also think we can add support for those after https://github.com/prefix-dev/shell/pull/148 is merged so that we'd take care of everything altogether. Let me know what you think! 
Also, in regards to the bash examples, if the separator is `{` or `}`, it shouldn't be that hard. We can look into those later since I don't think they are in priority.

**Pull request merged**